sympy/matrices/expressions/matexpr.py

```diff
@@ -1269,6 +1270,46 @@ def _make_matrix(x):
     return ImmutableDenseMatrix([[x]])
 
 
+class MatrixSet(Set):
+    """
+    MatrixSet represents the set of matrices with ``shape = (n, m)`` over the
+    given set.
+
+    Examples
+    ========
+
+    >>> from sympy.matrices import MatrixSet, Matrix
+    >>> from sympy import S, I
+    >>> M = MatrixSet(2, 2, set=S.Reals)
+    >>> X = Matrix([[1, 2], [3, 4]])
+    >>> X in M
+    True
+    >>> X = Matrix([[1, 2], [I, 4]])
+    >>> X in M
+    False
+
+    """
+    is_iterable = False
+    is_empty = False
+
+    def __new__(cls, n, m, set):
+        n, m, set = _sympify(n), _sympify(m), _sympify(set)
+        return Set.__new__(cls, n, m, set)
+
+    is_finite_set = property(lambda self: self.set.is_finite_set)
+    is_ComplexRegion = property(lambda self: self.set.is_ComplexRegion)
+
+    shape = property(lambda self: self.args[0:2])
+    set = property(lambda self: self.args[2])
```

Comment:
It's more usual to use `self.args[:2]` rather than `self.args[0:2]`.

I really don't like the ` = property(...)` style. That's not how decorators are normally used. There can be situations where it possibly makes sense but here I think it's just making the code confusing. It is generally better to stick to idiomatic Python code rather than try to save a few lines.